vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go

```diff
@@ -0,0 +1,267 @@
+	timeLayout       = "2006-01-02 15:04:05"
+	testTimestamp, _ = time.Parse(timeLayout, "2021-08-11 14:53:05")
+	testRequest      = model.Resources{
+		model.ResourceCPU:    model.CPUAmountFromCores(0.5),
+		model.ResourceMemory: model.MemoryAmountFromBytes(100),
+	}
+	testLimit = model.Resources{ // Setting limits to 3x of testRequest
+		model.ResourceCPU:    model.CPUAmountFromCores(1.0),
+		model.ResourceMemory: model.MemoryAmountFromBytes(300),
+	}
+	testUpdateModeAuto = vpa_types.UpdateModeAuto
+	testSelectorStr    = "label-1 = value-1"
+	testVpaID          = model.VpaID{
+		Namespace: testPodID1.Namespace,
+		VpaName:   "testVPA",
+	}
+)
+
+func addTestMemorySample(cluster *model.ClusterState, container model.ContainerID, memoryBytes float64, timeStamp time.Time) error {
+	sample := model.ContainerUsageSampleWithKey{
+		Container: container,
+		ContainerUsageSample: model.ContainerUsageSample{
+			MeasureStart: timeStamp,
+			Usage:        model.MemoryAmountFromBytes(memoryBytes),
+			Request:      testRequest[model.ResourceMemory],
+			Resource:     model.ResourceMemory,
+		},
+	}
+	return cluster.AddSample(&sample)
+}
+
+func addTestCPUHistogram(samples []float64, weight float64, testTimeStamp time.Time, options util.HistogramOptions) util.Histogram {
+	cpuHistogram := util.NewHistogram(options)
+	for _, sample := range samples {
+		cpuHistogram.AddSample(sample, weight, testTimestamp)
+	}
+
+	return cpuHistogram
+}
+
+func addTestMemoryHistogram(samples []float64, weight float64, timeStamp time.Time, options util.HistogramOptions) util.Histogram {
+	memoryHistogram := util.NewHistogram(options)
+	for _, sample := range samples {
+		memoryHistogram.AddSample(sample, weight, testTimestamp)
+	}
+
+	return memoryHistogram
+}
+
+func surgeSampleBasedMemoryEstimation(memorySample float64, maxAllowedMemory float64,
+	cpuHistogram util.Histogram, memoryPeaksHistogram util.Histogram,
+	cpuPercentile float64, memPercentile float64, fn getOOMParam) (model.Resources, int) {
+	surgeSamplesCount := 1
+	var resourceEstimation model.Resources
+	for {
+		if memorySample < maxAllowedMemory {
+			memorySample = fn(memorySample)
+		}
+		memoryPeaksHistogram.AddSample(memorySample, 1.0, testTimestamp)
+		resourceEstimation = GetPercentileResourceEstimation(cpuPercentile, memPercentile,
+			cpuHistogram, memoryPeaksHistogram)
+
+		if resourceEstimation[model.ResourceMemory] >= model.ResourceAmount(memorySample) {
+			break
+		}
+		surgeSamplesCount++
+	}
+
+	return resourceEstimation, surgeSamplesCount
+}
+
+func GetPercentileResourceEstimation(cpuPercentile float64, memPercentile float64,
+	cpuHistogram util.Histogram, memoryHistogram util.Histogram) model.Resources {
+
+	estimator := logic.NewPercentileEstimator(cpuPercentile, memPercentile)
+
+	return estimator.GetResourceEstimation(
+		&model.AggregateContainerState{
+			AggregateCPUUsage:    cpuHistogram,
+			AggregateMemoryPeaks: memoryHistogram,
+		})
+}
```

Comment:
Do we need this function? Can't we call [podResourceRecommender.GetRecommendedPodResources](https://github.com/gardener/autoscaler/blob/a842944a648b7e3d87b2915f4c5fea077db9969b/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L56:34) instead?

Reply:
ACK. Removed as not using estimator but using GetRecommendedPodResources now.